Spec: Avoid struct field conflicts in default values (#12841)
apache/iceberg: format/spec.md

The `initial-default` and `write-default` produce SQL default value behavior, without rewriting data files. SQL default value behavior when a field is added handles all existing rows as though the rows were written with the new field's default value. Default value changes may only affect future records and all known fields are written into data files. Omitting a known field when writing a data file is never allowed. The write default for a field must be written if a field is not supplied to a write. If the write default for a required field is not set, the writer must fail.

All columns of `unknown`, `variant`, `geometry`, and `geography` types must default to null. Non-null values for `initial-default` or `write-default` are invalid.

> Default values for the fields of a struct are tracked as `initial-default` and `write-default` at the field level. Default values for fields that are nested structs must not contain default values for the struct's fields (sub-fields). Sub-field defaults are tracked in sub-field's metadata. As a result, the default stored for a nested struct may be either null or a non-null struct with no field values. The actual default value is produced by setting each field default in a new struct.

Reviewer: non-null struct with no field values => empty struct?

```
The actual default value is produced by setting each field default in a new struct.
```

 Do we mean "New structs are produced by applying/combining all sub-field defaults."

Author: I think "non-null struct with no field values" is better because "empty" could be confusing in the context of a type. What does it mean to be empty? A struct with no field values means that there are no field-level values.

Maybe? I think that's less clear:
* "New structs are produced" is confusing because only the effective default value is produced this way
* "applying" or "combining" is subject to interpretation, but "setting" each field from its default is specific

Reviewer: I thought 'empty' meant a struct with no field values? When I read "with no field values" i'm a little lost because what does that mean that isn't empty?

"Effective default value is produced" sounds good to me, "actual" seems ambiguous.

> For example, a column `point` with fields `x` (default 0) and `y` (default 0) can be defaulted to `{"x": 0, "y": 0}` or `null`. The values stored for `initial-default` and `write-default` may be either `null` or an empty struct (`{}`) that indicates a non-null struct with field values set from each field's `initial-default` or `write-default`, respectively.

Reviewer: ```suggestion
For example, a  struct column `point` with sub-fields `x` (default 0) and `y` (default 0) can be defaulted to `{"x": 0, "y": 0}`. The values stored for `initial-default` and `write-default` would be an empty struct (`{}`) that indicates a non-null struct with field values set from each field's `initial-default` or `write-default`, respectively.
```

I am suggesting removing the "null" portion of the example and maybe placing that in a second example sentence.

"Alternatively, `point` could have `initial-default` and `write-default` set to null and any sub-field defaults would be ignored"
?

Author: I wouldn't use "sub-fields" because this is using "column" for the struct and fields for the struct's nested fields. I think that's clear enough.

Changing the sentence to use "would" seems to be more strict than I think it should be, but I can change to "A non-null default is stored by setting `initial-default` or `write-default` to an empty struct".

Reviewer: We went through some more examples on slack and I think some of those would be helpful here as well. 

IE Given initial defaults of .....
An null value would have an actual/effective value of ....
An empty struct would have a value of 
A struct with x defined would have a value of

Author: Added the table.

All columns of `unknown`, `variant`, `geometry`, and `geography` types must default to null. Non-null values for `initial-default` or `write-default` are invalid.

Default values for the fields of a struct are tracked as `initial-default` and `write-default` at the field level. Default values for fields that are nested structs must not contain default values for the struct's fields (sub-fields). Sub-field defaults are tracked in sub-field's metadata. As a result, the default stored for a nested struct may be either null or a non-null struct with no field values. The effective default value is produced by setting each fields' default in a new struct.

For example, a struct column `point` with fields `x` (default 0) and `y` (default 0) can be defaulted to `{"x": 0, "y": 0}` or `null`. A non-null default is stored by setting `initial-default` or `write-default` to an empty struct (`{}`) that will use field values set from each field's `initial-default` or `write-default`, respectively.

> | `point` default | `point.x` default | `point.y` default | Data value   | Result value |

Reviewer: This is really helpful. Thank you for adding it!

The `initial-default` and `write-default` produce SQL default value behavior, without rewriting data files. SQL default value behavior when a field is added handles all existing rows as though the rows were written with the new field's default value. Default value changes may only affect future records and all known fields are written into data files. Omitting a known field when writing a data file is never allowed. The write default for a field must be written if a field is not supplied to a write. If the write default for a required field is not set, the writer must fail.

All columns of `unknown`, `variant`, `geometry`, and `geography` types must default to null. Non-null values for `initial-default` or `write-default` are invalid.

> Default values for struct fields must be either null or non-null to avoid conflicts between the struct-level default for a field and a field-level default. Default values for struct fields are tracked at the field level.

Reviewer: ```suggestion
Default values for struct fields must be either null or non-null to avoid conflicts between the default for a struct and the default for a field within that struct. Default values for struct fields are tracked at the field level.
```

Reviewer 2: I'm a bit confused here. Is it `Default values for struct must be either null or non-null`?

Author: I'll update this. What I was trying to say is that the struct itself cannot set values for the fields, so there are only two cases: the default is null when the struct field is missing, or the struct is non-null and fields are assigned default values from field metadata.

Reviewer 3: I agree the wording in the PR can be read differently. We need to incorporate @rdblue's last comment, I think it is a lot more obvious.

Author: @aokolnychyi, there are two cases:
* The struct field's default is null -- engines should use null 
* The struct field's default is a non-null struct with no field values -- engines will be passed a struct-level default  that is produced by initializing a new struct and setting each field's default

* The `write-default` must be set when a field is added and may change
* When a required field is added, both defaults must be set to a non-null value
* When an optional field is added, the defaults may be null and should be explicitly set

> * When a field that is a struct type is added, its default may only be null or non-null. Default values for fields must be stored in field metadata.

Reviewer: We have a bit of a struct-field vs "field field" ambiguity for me here. 
```suggestion
* When a struct field is added, its default may only be null or non-null. Default values for fields within a struct must be stored in that field's metadata.
```

Reviewer 4: It might be just me, but as it stands, this sentence sounds odd:

> its default may only be null or non-null

I read this as: "this assertion may only be true or false". Which is necessarily true :-)

Imo we should define what we mean by `non-null` here. It seems that it denotes a non-null struct instantiated with default values for all its fields. The Go language as the term "zero-value" for that.

Author: The first change in the suggestion, adding a "struct field", is what I am trying to avoid. This requirement is stating what happens when a field is added and that field's type is a struct. Using "struct field" is not clear because all field additions are adding a struct field -- that is, a field within a struct. I'll reword to avoid some of the awkward language here.
